solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
     // but the addition of setLiveDocs means we needed to add volatile to "liveDocs".
     BitDocSet docs = liveDocs;
     if (docs == null) {
```

Comment:
Is there a race condition here when multiple match all docs queries come in at the same time? I think that before if there were multiple queries here at the same time then they would all sync on the same result in the filterCache. Hard to tell, this is tricky logic. Probably good to add a test or two?

Reply:
You're right that this case kind of undermines (for this special case) the improvements you introduced with SOLR-15555 ... I'll see what I can do about that...

Reply:
:+1: this should be addressed now, newly by lazy-initing `liveDocSet` via a single-element special-case CaffeineCache.